Stocks api implemented

kramarov-evg opened the pull request:
Closes #2, Closes #21

kramarov-evg committed:
Code clean-up

Changes:
- Tiny fixes to code style

kramarov-evg committed:
Created base for API communication

Changes:
- Defined data model
- Created API-communicator interface

kramarov-evg committed:
Created functionality for API communication

Changes:
- Created functions to retrieve data (w/caching)

kramarov-evg committed:
Non-working Commit

kramarov-evg committed:
Fixed null responses, added logger

kramarov-evg committed:
Finished work on API

- Created  custom period API data retrieval
- Modified disposables behaviour
- Simplified latest rate retrieval
- Added results caching

kramarov-evg committed:
XML fixed:

- No more adding extra ID's instead of just referencing

LeonidVolohov commented on app/src/main/java/com/stonks/api/stocks/StocksApiDataUtils.kt:
```diff
@@ -0,0 +1,205 @@
+        endDateTime: ZonedDateTime
+    ): Observable<StocksDataModel.RatesProcessed> {
+        val test = ZonedDateTime.of(2021, 1, 1, 10, 0, 0, 0, ZoneId.systemDefault())
+        Log.e("error", "${test >= startDateTime && test <= endDateTime}")
```
It is better to declare [TAG](https://developer.android.com/reference/android/util/Log) constant in your class like `val TAG = ClassName::class.java.name` instead of "error" string.  It turns out that you first indicate that the error belongs to the "error" (`Log.e(..., ...)`) level, and then in the tag parameter you indicate "error" string which is an unnecessary repetition, as it seems to me

kramarov-evg replied:
Nice, thx

LeonidVolohov commented on app/src/main/res/layout/fragment_stocks.xml:
```diff
@@ -49,22 +49,22 @@
         <Spinner
-            android:id="@+id/spinnerCurrency"
+            android:id="@+id/spinner_currency"
             android:layout_width="128dp"
```
All layout constants should be in `res/values` folder

kramarov-evg replied:
You mean `128dp` in this case?

LeonidVolohov reviewed (approved):
Have some small questions about logging, println and layout UI constants but overall everything is fine

owlerxan reviewed (approved):
It's ok

kramarov-evg committed:
Clean-up:

- Extracted size into dimensions
- Logging prettified
- Removed some redundant debugging code
